### ISSUE TITLE
Ansible raw fails randomly

### ISSUE THREAD

**Issue by duck-rh** [dw/mitogen#266]
Quack,

I'm using master commit 876a82f and Ansible 2.5.5.

This task:
```
    - name: Install Python
      raw: test -e /usr/bin/python || (apt -qqy update && apt install -qqy python-minimal)
      register: output
      changed_when: output.stdout != ""
```
sometimes, randomly, fails with:
```
 11280 1529065165.21643: _execute() done
 11280 1529065165.21658: dumping result to json
 11280 1529065165.21672: done dumping result, returning
 11280 1529065165.21751: done running TaskExecutor() for Jinta/TASK: Install Python [5c514f2f-680e-8e01-2ae9-00000000000d]
 11280 1529065165.21782: sending task result for task 5c514f2f-680e-8e01-2ae9-00000000000d
 11280 1529065165.21838: done sending task result for task 5c514f2f-680e-8e01-2ae9-00000000000d
 11280 1529065165.21851: WORKER PROCESS EXITING
fatal: [Jinta]: UNREACHABLE! => {"changed": false, "msg": "Connection timed out.", "unreachable": true}
```

Regards.
\\_o<

**Comment by chopraaa** [comment]
I've had the same issue with this.

**Comment by dw** [comment]
Thanks for reporting! It's almost certainly to do with the new async tasks implementation landed last week -- looking at this now

**Comment by dw** [comment]
Hi there,

Can you both please provide a little more info:

- The output of `-vvv` up to the point of a hang
- Number of targets in the run and the target OS
- OS of the host machine

**Comment by chopraaa** [comment]
> Presumably this raw task appears at the start of the run?

Yes, I'm using it to install python in my case.

I'll get back with the `-vvv` results.

**Comment by weeclemans** [comment]
I have the same issue (mitogen 0.2.3, ansible 2.7.5)

`$ cat temp.yml
---
- hosts:
    xenon.loc
  max_fail_percentage: 0
  any_errors_fatal: false
  gather_facts: false

  tasks:
    - raw: uname -a

$ ansible-playbook -vvv -i hosts temp.yml 
ansible-playbook 2.7.5
  config file = /home/homer/.ansible.cfg
  configured module search path = [u'/home/homer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 12 2018, 14:36:49) [GCC 5.4.0 20160609]
Using /home/homer/.ansible.cfg as config file
/home/homer/temp/work/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/home/homer/temp/work/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
Parsed /home/homer/temp/work/ansible/hosts inventory source with ini plugin
 ____________________
< PLAYBOOK: temp.yml >
 --------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

1 plays in temp.yml
 _______________________________
< PLAY [xenon.loc] >
 -------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

[pid 16739] 10:42:16.490631 D mitogen: mitogen.service.Pool(0x7fcc8b9ba350, size=16, th='MainThread'): initialized
[pid 16739] 10:42:16.491826 D ansible_mitogen.process: Service pool configured: size=16
META: ran handlers
 ____________
< TASK [raw] >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /home/homer/temp/work/ansible/temp.yml:9
[pid 16759] 10:42:16.532965 D ansible_mitogen.mixins: _low_level_execute_command(u'uname -a', in_data=<type 'NoneType'>, exe=False, dir=None)
[pid 16759] 10:42:16.533899 D mitogen: unix.connect(path='/tmp/mitogen_unix_Or7wa_')
[pid 16759] 10:42:16.534345 D mitogen: unix.connect(): local ID is 1, remote is 0
[pid 16739] 10:42:16.537385 D mitogen: mitogen.ssh.Stream(u'default').connect()
[pid 16739] 10:42:16.657276 D mitogen: create_child() child 16762 fd 60, parent 16739, cmd: ssh -o "LogLevel ERROR" -o "Compression yes" -o "ServerAliveInterval 15" -o "ServerAliveCountMax 3" -o "BatchMode yes" -o "StrictHostKeyChecking no" -o "UserKnownHostsFile /dev/null" -o "GlobalKnownHostsFile /dev/null" -C -o ControlMaster=auto -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no xenon.loc /usr/bin/python -c "'import codecs,os,sys;_=codecs.decode;exec(_(_(\"eNqFkV9rwjAUxZ/tp+hbEoyaKCgrBDZkjMHGoAx92GS0Jt2CNQnpv7lPv2sVbN3DnnJ/nHPvudzEdC1sMXbaKUwCT5sO6SwEyKzfYRIFA6hl5aaYUc4YuXBMu+RB5Sfe5rZQOO6C78K6Cw0ABBYHiM+TElL3oRAhkolvtEFhYmQrqm+1rcokzVUrT6rCT1JtJu5QflmDYM/BlW0o2sZa+UJb8xbNNm2sMrX2wOguflgxtBH9tpMHMMd9gfZxiPBel/ZTmajOk2Jn69vzO3pKSl1WUo3u59MZi/h8MbshiAQwuPG6VJhT9Pz4+sIYezcIFtpaCXcnwVJ84OPlpXXKwL2RTxEZe5VIzKcLzglFP9rBpMyJi29NUZOi42dk7hywbOvTga/czX/uv1vy3pa/nk6zSQ==\".encode(),\"base64\"),\"zip\"))'"
[pid 16739] 10:42:16.658604 D mitogen: mitogen.ssh.Stream(u'local.16762').connect(): child process stdin/stdout=60
[pid 16739] 10:42:16.950416 D mitogen: mitogen.ssh.Stream(u'local.16762'): received 'bash: /usr/bin/python: No such file or directory\n'
[pid 16739] 10:42:16.951416 D mitogen: mitogen.ssh.Stream(u'local.16762'): PID 16762 exited with return code 127
[pid 16759] 10:42:16.953269 D mitogen: mitogen.core.Stream(u'unix_listener.16739').on_disconnect()
[pid 16759] 10:42:16.953995 D mitogen: Waker(Broker(0x7fcc8b9b7f90) rfd=10, wfd=12).on_disconnect()
[pid 16739] 10:42:16.954033 D mitogen: mitogen.core.Stream(u'unix_client.16759').on_disconnect()
fatal: [xenon.loc]: UNREACHABLE! => {
    "changed": false, 
    "msg": "EOF on stream; last 300 bytes received: u'bash: /usr/bin/python: No such file or directory\\n'", 
    "unreachable": true
}
	to retry, use: --limit @/home/homer/temp/work/ansible/temp.retry
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

xenon.loc     : ok=0    changed=0    unreachable=1    failed=0   

[pid 16739] 10:42:16.986701 D mitogen: <mitogen.unix.Listener object at 0x7fcc8bb22f90>.on_disconnect()
[pid 16739] 10:42:16.987002 I mitogen: mitogen.service.Pool(0x7fcc8b9ba350, size=16, th='mitogen.service.Pool.7fcc8b9ba350.worker-1'): channel or latch closed, exitting: None
[pid 16739] 10:42:16.987128 D mitogen: Waker(Broker(0x7fcc8bb22a90) rfd=6, wfd=8).on_disconnect()
[pid 16739] 10:42:16.987356 I mitogen: mitogen.service.Pool(0x7fcc8b9ba350, size=16, th='mitogen.service.Pool.7fcc8b9ba350.worker-2'): channel or latch closed, exitting: None
[pid 16739] 10:42:16.987425 I mitogen: mitogen.service.Pool(0x7fcc8b9ba350, size=16, th='mitogen.service.Pool.7fcc8b9ba350.worker-3'): channel or latch closed, exitting: None
`

**Comment by weeclemans** [comment]
I think so - Need functionality to disable mitogen on RAw tasks ...

**Comment by dw** [comment]
[Note the raw module currently requires Python](https://mitogen.readthedocs.io/en/dmw/changelog.html#mitogen-for-ansible) to be installed. Fixing that is basically blocked on #419 to avoid any more spaghetti code.

But that doesn't seem to explain your situation. What kind of machine is it connecting to and how many machines are in a typical run?

We had some super scary bugs fixed over the past 6 months -- including several where FDs could be closed at random, this might explain it. At one point bootstrap could fail if the machine was low on RAM /and/ user was SSHing into an unprivileged container.

Another avenue is a difference in how Mitogen interprets 'connected' compared to Ansible -- it requires everything to happen up to and including the remote interpreter saying hello before it is marked connected. Is there some chance that the 'python' command is failing on that machine, and somehow the SSH connection is otherwise being held open? For example, a crazy bashrc that backgrounds some process might do this

I'm going to have a play with raw specifically, but just in case you can tickle it again, [this strace wrapper technique](https://mitogen.readthedocs.io/en/dmw/ansible.html#post-authentication-bootstrap-failure) is very effective at revealing startup problems that aren't so easy to capture in a log.

**Comment by dsgnr** [comment]
I can confirm this is an issue when Mitogen is installed on both Ubuntu 16.04 and Ubuntu 18.04. 
```- hosts: all
  gather_facts: False
  
  tasks:
  - name: install python 2
    raw: test -e /usr/bin/python || (apt -y update && apt install -y python-minimal)```
Simply does not work currently with latest Mitogen release (I'm using Ansible 2.7.2).

Commenting Mitogen on the ansible.cfg file runs the task fine

**Comment by dw** [comment]
@dsgnr I think that's a separate issue to ducks'.. Mitogen does not have any raw support right now (it's coming soon)

You can achieve the same effect (if convenient) by switching that one task to the linear strategy within the same file. See the third bullet point on https://mitogen.readthedocs.io/en/latest/changelog.html#known-issues

**Comment by dsgnr** [comment]
Thanks for that @dw. I hadn’t spotted that as a known issue!

**Comment by dw** [comment]
Raw is a real easy feature to do, but it basically requires cutting a new dev branch, that's why it hasn't been done yet. I'm trying to find a new way to do all this in 0.2 without making a mess, might be better for long term stability than risking a dev branch where all hell breaks loose ;)

**Comment by dberardo-com** [comment]
still now news on this ?

**Comment by dberardo-com** [comment]
the problem is in air gapped environments, the ansible hosts i am controlling have no internet access to python cant be easily installed

**Comment by duck-rh** [comment]
I had to let go of Mitogen at some point but I'm still interested. I certainly won't bother you if you prefer to close this ticket but I think in this case the documentation should make this limitation clear (with the workaround).

That said I'm still using the snippet above, updated for Python 3. I remember that with CentOS/RHEL having a system-python separate from the usual python package I wanted to make sure I got the latter. Also if you have no access to VM preseeding (to install the package at this stage), that can be useful; you're not always in charge of all the pieces of infra.

**Comment by dberardo-com** [comment]
also: how can strategy be switched to linear for just a single task ? does it require a new play ?

**Comment by baryluk** [comment]
@dberardo-com I see, yeah, I am not sure if this can be applied to individual task.

**Comment by moreati** [comment]
Ansible strategy can only be changed per play. Making it per task isn't something Mitogen could do.
